### PR TITLE
feat(sidebar): Redesign sidebar toggle and mobile layout

### PR DESCRIPTION
**Pull request by gadenbuie**
## TODO

- [x] Remove top padding on always open sidebars
- [x] Always open sidebars in mobile go to "flow" layout
- [x] Use `arrow-bar-left` for the toggle icon
- [x] Should we use circle shape to match full screen card toggle?
    - Keeping current radius, but there's a `--bslib-collapse-toggle-border-radius` that can be used to set this.
- [x] Default to `sidebar(position = "right")`? 
	- (Not now)
- [x] Deprecate or revisit sidebar mobile height shenanigans?
	- `max_height_mobile` now only applies to `open = "always"` sidebars and is ignored otherwise. A warning is provided if `max_height_mobile` is provided with another `open` value.

## Intro

This PR revisits the sidebar toggle with a new design and some updated behavior. At a high level:

* Toggle moves to the top of the sidebar. When collapsed, the toggle is outside the sidebar area and when expanded the toggle moves into the sidebar area.

* To accomodate this, we hold space in two places: 
    * Add padding at the top of every sidebar to accomodate the toggle button
    * When collapsed, we also increase the padding in the main area adjacent to the sidebar

* The colors of the main content area are now controlled via CSS variables, `--bslib-sidebar-main-fg` and `--bslib-sidebar-main-bg`. I think this is a better pattern anyway, but it allows the toggle icon to be colored according to the foreground color of the region where its placed. (Previously it was static and only used the sidebar fg/bg colors.)

* Mobile view is completely revisited, ultimately it's much simpler. The sidebar operates in the same way (expands from the side), but takes up the complete layout area as an overlay over the main content area. Only one sidebar can be expanded at a time, so nested sidebars in the mobile view are simplified.

## Demo App

Here's the code for a demo app to test features, appearance, etc. None of the controls are wired up, they're just there to help motivate appearance discussions.

<details>
<summary>Demo app with bunches of sidebars</summary>

```r
library(shiny)
pkgload::load_all()

library(htmltools)
library(plotly)
library(leaflet)

plotly_widget <-
  plot_ly(x = diamonds$cut) |>
  config(displayModeBar = FALSE) |>
  layout(margin = list(t = 0, b = 0, l = 0, r = 0))

leaflet_widget <-
  leaflet(
    options = leafletOptions(attributionControl = FALSE)
  ) |>
  setView(lng = 174.768, lat = -36.852, zoom = 5) |>
  addMarkers(lng = 174.768, lat = -36.852, popup = "The birthplace of R") |>
  addTiles()

accordions <- function(id) {
  ns <- shiny::NS(id)
  accordion(
    open = FALSE,
    accordion_panel(
      "Dropdowns",
      value = ns("Numerical"),
      icon = bsicons::bs_icon("menu-app"),
      selectInput(ns("cut"), "Cut", choices = LETTERS),
      selectInput(ns("clarity"), "Clarity", choices = LETTERS)
    ),
    accordion_panel(
      "Numerical",
      value = ns("Dropdowns"),
      icon = bsicons::bs_icon("sliders"),
      sliderInput(ns("depth"), "Depth", 0, 100, 25),
      sliderInput(ns("table"), "Table", 0, 100, 50)
    )
  )
}

ui <- page_navbar(
  title = "Sidebar Refresh",
  fillable = TRUE,
  nav_spacer(),
  nav_panel(
    "Sidebar Page",
    layout_sidebar(
      sidebar = sidebar(
        title = "Diamond in the rough",
        actionButton("btn2", "Do the thing"),
        selectInput("select", "Select the thing", choices = LETTERS),
        textInput("text", "Type the thing", value = "Hello, world!")
      ),
      card(
        height = 300,
        full_screen = TRUE,
        card_header("A sidebar layout inside a card"),
        layout_sidebar(
          fillable = TRUE,
          sidebar = sidebar(
            actionButton("btn", "A button"),
            open = FALSE
          ),
          plotly_widget
        )
      ),
      card(
        full_screen = TRUE,
        class = "bg-dark",
        card_header("A filling map"),
        layout_sidebar(
          class = "p-0",
          fg = "black",
          sidebar = sidebar(
            title = "The Birthplace of R",
            fg = "white",
            p(
              "Did you know that R was born in New Zealand?",
              "Zoom in to find out exactly where."
            ),
            position = "right",
            open = FALSE
          ),
          card_body(class = "p-0", leaflet_widget)
        )
      )
    )
  ),
  nav_panel(
    "Sidebar Accordion",
    card(
      card_header("Groups of diamond filters"),
      layout_sidebar(
        sidebar = sidebar(accordions("one")),
        plotly_widget
      )
    ),
    card(
      card_header("Groups of diamond filters"),
      layout_sidebar(
        sidebar = sidebar(title = "Diamond studded accordions", accordions("two")),
        plotly_widget
      )
    )
  ),
  nav_panel(
    "Nested Sidebars",
    card(
      card_header("Nested sidebars"),
      layout_sidebar(
        fillable = TRUE,
        class = "p-0",
        sidebar = sidebar(
          width = "20%",
          bg = "#FF7847",
          fg = "",
          "Left Sidebar 1"
        ),
        layout_sidebar(
          fillable = TRUE,
          class = "p-0",
          border = FALSE,
          border_radius = FALSE,
          sidebar = sidebar(
            width = "30%",
            bg = "#FF70A6",
            position = "right",
            "Right Sidebar"
          ),
          layout_sidebar(
            border = FALSE,
            border_radius = FALSE,
            sidebar = sidebar(
              width = "40%",
              bg = "#FFB499",
              "Left Sidebar 2",
            ),
            lorem::ipsum(3, 1)
          )
        )
      )
    ),
    card(
      card_header("Nested sidebars (right)"),
      layout_sidebar(
        fillable = TRUE,
        class = "p-0",
        sidebar = sidebar(
          width = "20%",
          bg = "#FF7847",
          fg = "",
          "Left Sidebar"
        ),
        layout_sidebar(
          fillable = TRUE,
          class = "py-0 pe-0",
          border = FALSE,
          border_radius = FALSE,
          sidebar = sidebar(
            width = "30%",
            bg = "#FF70A6",
            position = "right",
            "Right Sidebar 1"
          ),
          layout_sidebar(
            border = FALSE,
            border_radius = FALSE,
            sidebar = sidebar(
              width = "40%",
              bg = "#FFB499",
              position = "right",
              "Right Sidebar 2",
            ),
            lorem::ipsum(3, 1)
          )
        )
      )
    )
  ),
  nav_panel(
    "Static",
    card(
      card_header("Static sidebar"),
      layout_sidebar(
        sidebar = sidebar(
          title = "Static sidebar",
          "This sidebar is always open",
          open = "always"
        ),
        lorem::ipsum(3, 4)
      )
    ),
    card(
      card_header("Static sidebar (right)"),
      layout_sidebar(
        sidebar = sidebar(
          title = "Static sidebar",
          "This sidebar is always open",
          position = "right",
          open = "always"
        ),
        lorem::ipsum(3, 4)
      )
    )
  )
)

server <- function(input, output, session) {

}

shinyApp(ui, server)
```

</details>

### CODE REVIEW

File: R/sidebar.R [hunk 1]
```diff
@@ -291,10 +291,10 @@ sidebar_toggle <- toggle_sidebar
 
 collapse_icon <- function() {
   if (!is_installed("bsicons")) {
-    icon <- "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\" class=\"bi bi-chevron-down collapse-icon\" style=\"fill:currentColor;\" aria-hidden=\"true\" role=\"img\" ><path fill-rule=\"evenodd\" d=\"M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z\"></path></svg>"
+    icon <- "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\" class=\"bi bi-arrow-bar-left collapse-icon\" style=\"height:;width:;fill:currentColor;\" aria-hidden=\"true\" role=\"img\" ><path fill-rule=\"evenodd\" d=\"M12.5 15a.5.5 0 0 1-.5-.5v-13a.5.5 0 0 1 1 0v13a.5.5 0 0 1-.5.5ZM10 8a.5.5 0 0 1-.5.5H3.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L3.707 7.5H9.5a.5.5 0 0 1 .5.5Z\"></path></svg>"
```
Comment (cpsievert) [inline R/sidebar.R:294]:
```suggestion
    icon <- "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\" class=\"bi bi-arrow-bar-left collapse-icon\" style=\"fill:currentColor;\" aria-hidden=\"true\" role=\"img\" ><path fill-rule=\"evenodd\" d=\"M12.5 15a.5.5 0 0 1-.5-.5v-13a.5.5 0 0 1 1 0v13a.5.5 0 0 1-.5.5ZM10 8a.5.5 0 0 1-.5.5H3.707l2.147 2.146a.5.5 0 0 1-.708.708l-3-3a.5.5 0 0 1 0-.708l3-3a.5.5 0 1 1 .708.708L3.707 7.5H9.5a.5.5 0 0 1 .5.5Z\"></path></svg>"
```

File: inst/components/scss/sidebar.scss [hunk 2]
```diff
@@ -54,14 +64,17 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
     border-right: var(--bslib-sidebar-vert-border);
     border-top-right-radius: 0;
     border-bottom-right-radius: 0;
-    background-color: var(--bslib-sidebar-bg);
     color: var(--bslib-sidebar-fg);
+    background-color: var(--bslib-sidebar-bg);
+    // TODO: background can be transparent to be dark mode adaptive, need opaque "subtle" variant
+    backdrop-filter: blur(5px);
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:70]: Is this still needed now that we have the `opacity: 0` thing?

```suggestion
```

File: inst/components/scss/sidebar.scss [hunk 3]
```diff
@@ -54,14 +64,17 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
     border-right: var(--bslib-sidebar-vert-border);
     border-top-right-radius: 0;
     border-bottom-right-radius: 0;
-    background-color: var(--bslib-sidebar-bg);
     color: var(--bslib-sidebar-fg);
+    background-color: var(--bslib-sidebar-bg);
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:68]: For sake of a minimal diff

```suggestion
    background-color: var(--bslib-sidebar-bg);
    color: var(--bslib-sidebar-fg);
```

File: inst/components/scss/sidebar.scss [hunk 4]
```diff
@@ -1,20 +1,27 @@
...
   --bslib-sidebar-fg: #{$bslib-sidebar-fg};
+  --bslib-sidebar-main-fg: var(--bs-card-color, var(--bs-body-color));
+  --bslib-sidebar-main-bg: var(--bs-card-bg, var(--bs-body-bg));
+  --bslib-sidebar-toggle-bg: #{$bslib-sidebar-toggle-bg};
   --bslib-sidebar-padding: calc(var(--bslib-spacer) * 1.5);
   --bslib-sidebar-icon-size: var(--bslib-spacer, 1rem);
-  --bslib-collapse-toggle-border: #{$bslib-sidebar-border};
-  --bslib-collapse-toggle-transform: 90deg;
-  --bslib-collapse-toggle-right-transform: -90deg;
+  // --bslib-collapse-toggle-border: ;
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:20]:
```suggestion
```

File: inst/components/scss/sidebar.scss [hunk 5]
```diff
@@ -118,30 +120,49 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
     }
   }
 
+  &[data-bslib-sidebar-open="always"] {
+    > .sidebar > .sidebar-content {
+      // remove toggle button padding adjustment if there isn't a button
+      padding-top: var(--bslib-sidebar-padding);
+    }
+  }
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:128]: This way we wouldn't need this logic in two places?

```suggestion
  &:not([data-bslib-sidebar-open="always"]) {
    > .sidebar > .sidebar-content {
      // remove toggle button padding adjustment if there isn't a button
      padding-top: calc(var(--bslib-sidebar-padding) * 2);
    }
  }
```

File: inst/components/scss/sidebar.scss [hunk 6]
```diff
@@ -54,14 +64,17 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
...
+    background-color: var(--bslib-sidebar-bg);
+    // TODO: background can be transparent to be dark mode adaptive, need opaque "subtle" variant
+    backdrop-filter: blur(5px);
 
     > .sidebar-content {
       display: flex;
       flex-direction: column;
       gap: var(--bslib-spacer, 1rem);
       padding: var(--bslib-sidebar-padding);
+      padding-top: calc(var(--bslib-sidebar-padding, 1rem) * 2);
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:77]: See suggestion below

```suggestion
```

File: inst/components/scss/sidebar.scss (hunk 3 above)
Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:68]: If we drop `backdrop-filter` then yeah, I'll swap this back

File: inst/components/scss/sidebar.scss (hunk 4 above)
Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:20]: I think we should keep this as one step below actual documentation

File: inst/components/scss/sidebar.scss (hunk 5 above)
Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:128]: Yeah good point (note to self the change should be higher up closer to where `padding-top` is modified currently)

File: inst/components/scss/sidebar.scss [hunk 7]
```diff
@@ -236,89 +269,75 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
 
 @include media-breakpoint-down(sm) {
   .bslib-sidebar-layout, .bslib-sidebar-layout.sidebar-right {
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:271]: Kind seems like you could get rid of this and move the CSS below it into the `.bslib-sidebar-layout {}` below it? 

```suggestion
```

File: srcts/src/components/sidebar.ts [hunk 8]
```diff
@@ -93,6 +93,9 @@ class Sidebar {
       ":scope > .sidebar-content > .accordion"
     );
     if (sideAccordion) {
+      this.layout.sidebar
+        .querySelector(".sidebar-content")
+        ?.classList?.add("has-accordion");
```
Comment (cpsievert) [inline srcts/src/components/sidebar.ts:98]:
```suggestion
    if (sideAccordion) {
      sideAccordion.parentElement.classList.add("has-accordion");
```

File: inst/components/scss/sidebar.scss [hunk 9]
```diff
@@ -118,30 +120,49 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
...
+    border: none;
+    border-radius: var(--bslib-collapse-toggle-border-radius);
+    height: calc(var(--bslib-sidebar-icon-size) * 2);
+    width: calc(var(--bslib-sidebar-icon-size) * 2);
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    padding: 0;
     color: var(--bslib-sidebar-fg);
+    background-color: unset;
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:147]: Why is this needed?

File: inst/components/scss/sidebar.scss (hunk 2 above)
Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:70]: I decided to keep this, it looks better with the blur, even as the contents fade out.

File: inst/components/scss/sidebar.scss (hunk 5 above)
Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:128]: Oh wait, I was wrong, it has to be here because the `data-bslib-sidebar-open` attribute appears on the layout element and the `.sidebar-content` class is nested too deeply above.

I'm torn about taking this change and I am leaning toward not taking it. I generally would prefer to use the positive state rather than the negation, i.e. here we know specifically that the `"always"` sidebars don't want the extra padding. I think this CSS will be easier to maintain in the long run if the rule matches the nature of the exception.

File: inst/components/scss/sidebar.scss [hunk 10]
```diff
@@ -54,14 +64,16 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
...
     color: var(--bslib-sidebar-fg);
+    background-color: var(--bslib-sidebar-bg);
+    backdrop-filter: blur(5px);
 
     > .sidebar-content {
       display: flex;
       flex-direction: column;
       gap: var(--bslib-spacer, 1rem);
       padding: var(--bslib-sidebar-padding);
+      padding-top: calc(var(--bslib-sidebar-padding, 1rem) * 2);
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:76]: Explain that this extra padding is for the toggle and that we undo it when there is no toggle

File: inst/components/scss/sidebar.scss [hunk 11]
```diff
@@ -235,90 +267,80 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
...
-    // required by sidebar init js when `sidebar(open = "desktop")`
-    &[data-bslib-sidebar-open="desktop"] {
-      --bslib-sidebar-js-init-collapsed: true;
+  .bslib-sidebar-layout {
+    // Tell sidebar init js we're on mobile for `sidebar(open = "desktop")`
+    &, &.sidebar-right {
+      &[data-bslib-sidebar-open="desktop"] {
+        --bslib-sidebar-js-init-collapsed: true;
+      }
     }
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:276]:
```suggestion
    &[data-bslib-sidebar-open="desktop"] {
      --bslib-sidebar-js-init-collapsed: true;
    }
```

File: inst/components/scss/sidebar.scss [hunk 12]
```diff
@@ -235,90 +267,80 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
...
-      grid-row: 1 / 2;
-      grid-column: 1 / 2;
-      width: 100%;
-      border: none;
-      border-bottom: var(--bslib-sidebar-horiz-border);
-      border-radius: 0;
+    // Remove left or right edge sidebar borders in mobile view
+    &:not(.sidebar-right) {
+      > .sidebar { border-right: none }
     }
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:281]:
```suggestion
    > .sidebar {  border: none; }
```

File: inst/components/scss/sidebar.scss [hunk 13]
```diff
@@ -235,90 +267,80 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--bslib-sidebar-icon-size)),
...
-    > .main {
-      grid-row: 2 / 3;
-      grid-column: 1 / 2;
-      border-top-left-radius: 0;
-      border-top-right-radius: 0;
-      border-bottom-right-radius: inherit;
-      border-bottom-left-radius: inherit;
+    &.sidebar-right {
+      > .sidebar { border-left: none }
     }
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:284]: Could all of this be just `.sidebar{border:none}`? (but we might need `border-bottom` for the alywas open?)

Comment (gadenbuie) [reply inst/components/scss/sidebar.scss]: Yes, done in 020eb11e906602ab3145221c85122397a4344c77 and I added the border-top for always open sidebars

File: inst/components/scss/sidebar.scss (hunk 11 above)
Comment (gadenbuie) [reply inst/components/scss/sidebar.scss]: yes, done in 020eb11e906602ab3145221c85122397a4344c77

**Review by cpsievert — APPROVED** [review]
Nicely done! :shipit:

**Comment by daattali** [comment]
Don't feel an obligation to respond if it's a complex answer, but I'm curious: why did the sidebar toggle change? Without knowing the context of this change, the new toggle button seems to me less ideal, because now it (1) takes up precious vertical real estate from the side bar which makes the top of the sidebar look empty, and (2) takes horizontal real estate from the main content when the sidebar is collapsed, which similarly makes the main content are look a bit awkward by having the left side be non symmetric with the right side.

**Comment by gadenbuie** [comment]
Hi @daattali! The changes were motivated by a few things all coming together. Above all, we wanted to orient the toggle at the top, rather than the bottom, and to have the sidebar expand to fill the full container width on mobile. Those both motivated the change in position and the new behavior where the toggle floats into and out of the sidebar. We too noted the symmetry issues and are planning on addressing that before release.

We also felt that the previous design was overly compact and the usability suffered. The new toggle has a much bigger touch target. Along these lines, a common design thread in our latest work is to create more breathing room and space in default Shiny apps.

We appreciate the feedback! If you're able to share specific examples from your work, either as screenshots or reprex apps, we'd love to take a look and take them into consideration.

**Comment by gadenbuie** [comment]
It does look like there's a bug in those screenshots. The left gutter between the sidebar and the content is not constant, it's increased when the sidebar is collapsed, but returns to normal when the sidebar is open. Either the screenshots are missing the right edge of the app or there's a bug in the (version used for the) screenshots. Regardless, we'll be addressing the asymmetry in a follow-up PR soon.
